### DISCUSSION TITLE
Layout library for Python - built with Taffy

### DISCUSSION THREAD

**Discussion by mortencombat** [DioxusLabs/taffy#587] [Show and tell]
Hi all,

First of all, I'd like to thank you for your efforts in building and continuously improving on Taffy, it's highly appreciated 👍  

I'm the creator and maintainer of [stretchable](https://github.com/mortencombat/stretchable), a layout library for Python that is based on Taffy. Version 1.0 was released yesterday, it's still early days and there's certainly room for improvement, but it should be quite useable for those who are working in the Python ecosystem 😄

**Comment by alice-i-cecile** [comment]
Neat! I'm really happy to see it being used elsewhere. How was the interop process?

**Reply by mortencombat** [reply]
I think it was quite neat with PyO3. The biggest hurdle for me was learning Rust (at a basic level), of which I had no prior experience, especially coming from something like Python 😄

**Comment by nicoburns** [comment]
That's awesome. We should probably add a "bindings" section to the readme and link out to stretchable and other bindings as/when they are created.

FYI, a 0.4 release of Taffy is imminent which will come with support for block layout, the overflow property / scrollable layouts, and also reworks how measure functions are done to be more general and allow borrowed context to be accessed during layout (you'll probably need to rework your integration a little if you upgrade, but they should be internal changes and shouldn't require any changes to the API of your library).

Also, we could potentially coordinate on testing. I'm vaguely planning to generalise Taffy's test generator to be able to generate tests for multiple languages in order to accommodate our own first-party C and WASM bindings (or possibly adapt [Yoga's test generator](https://github.com/facebook/yoga/tree/main/gentest) as we mostly share a test suite with Yoga and Yoga's generator already supports multiple backends - some Yoga's generator is in Ruby though, which I am keen to avoid! So potentially a port of Yoga's generator to either Rust or JavaScript)

**Reply by mortencombat** [reply]
Please do, I’d be happy to see some people start making use of it.

That’s some nice additions and improvements for 0.4, I’ll definitely be looking to upgrade stretchable to support that when it’s ready.

It would be quite nice to be able to use the test generator directly. At the moment I’m parsing the HTML fixtures into nodes/styles, the style parsing is a bit cumbersome, on the other hand it will be quite useful for how I’ll be using stretchable upstream, and it is one of the areas where I’ll be working to improve on stretchable.
